apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/TransactionPool.cs

```diff
@@ -0,0 +1,247 @@
+        static readonly ConcurrentDictionary<ByteString, Session> s_activeTransactionTable = new ConcurrentDictionary<ByteString, Session>();
+
+        /// <summary>
+        /// Creates a transaction with the given transactoin options.
+        /// Depending on the state of the session, this may or may not result in a request
+        /// to Spanner over the network.
+        /// </summary>
+        /// <param name="session"></param>
+        /// <param name="client"></param>
+        /// <param name="options"></param>
+        /// <returns></returns>
+        public static async Task<Transaction> BeginPooledTransactionAsync(this SpannerClient client, Session session,
+            TransactionOptions options)
+        {
+            var info = s_sessionInfoTable.GetOrAdd(session, s => new SessionInfo {SpannerClient = client});
+        
+            //we need to await for previous task completion anyway -- otherwise there is a bad race condition.
+            await info.WaitForPreWarmAsync().ConfigureAwait(false);
+            
+            //now we see if we can return the prewarmed transaction.
+            // a lot has to be correct for us to return the precreated transaction.
+            // we have to have a valid transaction object.
+            // it needs to exist in the activetransactiontable to indicate that it hasn't been closed
+            // the options on it must equal the passed in options.
+            if (info.ActiveTransaction != null
+                && !info.ActiveTransaction.Id.IsEmpty
+                && s_activeTransactionTable.ContainsKey(info.ActiveTransaction.Id))
+            {
+                if (info.ActiveTransactionOptions != null
+                    && info.ActiveTransactionOptions.Equals(options))
+                {
+                    return info.ActiveTransaction;
+                }
+                //cache hit, but no match on options
+                Session ignored;
+                s_activeTransactionTable.TryRemove(info.ActiveTransaction.Id, out ignored);
+            }
+
+            //ok, our cache hit didnt work for whatever reason.  Let's create a transaction with the given options and return it.
+            await info.CreateTransactionAsync(session, options);
+            if (info.ActiveTransaction != null)
+            {
+                s_activeTransactionTable.AddOrUpdate(info.ActiveTransaction.Id, session, (id, s) => session);
+            }
+            return info.ActiveTransaction;
+        }
+
+        /// <summary>
+        /// Sets the transaction option on the session to be an implicit transaction.
+        /// This option is often used when a session is used for reads without specifying a transaction.
+        /// </summary>
+        /// <param name="session"></param>
+        /// <returns></returns>
+        public static Task SetImplicitTransactionAsync(this Session session)
+        {
+            return RemoveFromTransactionPoolAsync(session);
```

Comment:
It improves the readability from the layer above because its communicating intent.  The fact that the pool implements the behavior by doing the remove is an implementation detail that the caller should not know.

Someone calling this wants to say that "This session is using implicit transactions, please update its internal cached state as such".